Add `store_id` to WC_Tracker and Tracks

abelsonlive opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

This PR adds a unique `store_id` to WC Tracker snapshots and Tracks events. This ID is meant to aid the process of tracking merchants over time even if they change their store's URL. For now, **this change does not handle the case when an admin may copy an installations database to create a new store**.  We'll attempt to address that edge case in a future PR. 

This PR also adds the Jetpack `blog_id` to WC Tracker snapshots which is already included in Tracks events.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

There are three separate pathways for testing:

#### Testing the addition of the `store_id` to WC Tracker

1. Setup a local testing environment, for instance via `pnpm -- wp-env start`
2. Execute the `wp-cli` command `wc tracker snapshot`: `pnpm -- wp-env run cli wp wc tracker snapshot --format=json`. You should see a uuid4-formatted `store_id` in the output. You can use [`jq`](https://jqlang.github.io/jq/) to more quickly access this key: `pnpm -- wp-env run cli wp wc tracker snapshot --format=json | jq '.[0].store_id'`


#### Testing the addition of the `store_id` to Tracks

1. Start a local environment, eg: `pnpm -- wp-env start`
2. Enable WooCommerce tracking in the admin by going to `WooCommerce > Settings > Advanced > WooCommerce.com` and clicking the checkbox for `Allow usage of WooCommerce to be tracked`.
3. Open your browser's network tab and add a filter for requests that include `t.gif`.
4. Click around the WooCommerce admin (for instance by navigating to `WooCommerce > Home` and you should see some Tracks events being transmitted with the `store_id` in the payload:

<img width="512" alt="Screenshot 2023-10-11 at 3 30 46 PM" src="https://github.com/woocommerce/woocommerce/assets/1621721/cec18a93-6991-4ef3-9365-a181c87b5d45">

#### Testing the addition of the Jetpack `blog_id` to WC Tracker

1. Start an environment via Jurassic Ninja using this branch (`add/store-id-to-wc-tracker-and-tracks`) and make sure Jetpack is installed.
2. Once the environment is up, go through the store setup process ensuring that you've enabled WooCommerce tracking. Alternatively, enable WooCommerce tracking in the admin by going to `WooCommerce > Settings > Advanced > WooCommerce.com` and clicking the checkbox for `Allow usage of WooCommerce to be tracked`.
3. Connect Jetpack to your WPCom account. 
4. Use the ssh connection details in Jurassic Ninja to connect to its server and navigate to the specified `public` directory.
5. Run `wp wc tracker snapshot --format=json` and confirm that the `blog_id` is present in the output.

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message

Add a `store_id` to WC Tracker snapshots and Tracks events.

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

abelsonlive committed:
add store_id to WC_Tracker and Tracks; add blog_id to WC_Tracker

abelsonlive committed:
update wc-install-test; add wc-tracks-test

abelsonlive committed:
improved blog_id accessor and more tests

abelsonlive committed:
attempt to mock

abelsonlive committed:
spacing

abelsonlive committed:
revert jetpack options changes

abelsonlive committed:
linting fixes

abelsonlive committed:
linting fixes

abelsonlive committed:
rm blog_id test

abelsonlive commented on plugins/woocommerce/src/Admin/Features/ProductBlockEditor/ProductTemplates/SimpleProductTemplate.php:
```diff
@@ -564,15 +564,15 @@ private function add_inventory_group_blocks() {
+					'property'     => 'manage_stock',
+					'disabled'     => 'yes' !== get_option( 'woocommerce_manage_stock' ),
 					'disabledCopy' => sprintf(
 						/* translators: %1$s: Learn more link opening tag. %2$s: Learn more link closing tag.*/
-							__( 'Per your %1$sstore settings%2$s, inventory management is <strong>disabled</strong>.', 'woocommerce' ),
-							'<a href="' . admin_url( 'admin.php?page=wc-settings&tab=products&section=inventory' ) . '" target="_blank" rel="noreferrer">',
-							'</a>'
-						),
+						__( 'Per your %1$sstore settings%2$s, inventory management is <strong>disabled</strong>.', 'woocommerce' ),
+						'<a href="' . admin_url( 'admin.php?page=wc-settings&tab=products&section=inventory' ) . '" target="_blank" rel="noreferrer">',
+						'</a>'
+					),
```
linting changes.

abelsonlive commented on plugins/woocommerce/src/Admin/Features/ProductBlockEditor/ProductTemplates/SimpleProductTemplate.php:
```diff
@@ -782,7 +782,7 @@ private function add_variation_group_blocks() {
 			return;
 		}
-		$variation_fields = $variation_group->add_block(
+		$variation_fields          = $variation_group->add_block(
```
linting change.

abelsonlive committed:
remove blog_id tests

abelsonlive commented on plugins/woocommerce/includes/class-wc-install.php:
```diff
@@ -610,6 +618,20 @@ private static function maybe_update_db_version() {
+	 *
+	 * @since 8.2.0
+	 */
+	public static function maybe_set_store_id() {
+		$store_id = get_option( self::STORE_ID_OPTION, false );
+		if ( ! $store_id ) {
+			$store_id = wp_generate_uuid4();
+			add_option( self::STORE_ID_OPTION, $store_id );
+		}
+		return $store_id;
```
This is a public method to enable unit testing.

nerrad replied:
I'm assuming the only reason why this is returning the store id is because it's being used in unit testing since everywhere in the business logic I see it being retrieved via `get_option`. 

At a glance, I don't think unit tests should require returning the store_id here (or that the method needs to be public). I'd prefer to see this remain private and also act as a setter (no need to return the set value).

abelsonlive replied:
Yes, that's a good point! I will change it to not return the `store_id` and update the tests accordingly. 

However, I'm not sure how to make it a `private` method and still access it within the test. It seems that many other methods in this class are `public` to enable unit testing, for instance [`public_row_meta`](https://github.com/woocommerce/woocommerce/blob/eda463d542a3206f4068ddbcdaab60cb81754838/plugins/woocommerce/tests/php/includes/class-wc-install-test.php#L90). Is there a pattern I can follow to enable to testing while making this a `private` method? Forgive me as I'm a noob when it comes to PHP/WP development 😭

abelsonlive commented on plugins/woocommerce/includes/class-wc-tracker.php:
```diff
@@ -133,7 +133,10 @@ public static function get_tracking_data() {
-		$data['url'] = home_url();
+		$data['url']      = home_url();
+		$data['store_id'] = get_option( 'woocommerce_store_id', null );
+		$data['blog_id']  = class_exists( 'Jetpack_Options' ) ? Jetpack_Options::get_option( 'id' ) : null;
```
We're adding the Jetpack `blog_id` here to make WC Tracker snapshots consistent with Tracks events.

nerrad replied:
This might be a pedantic question but are you sure `get_option` is always available on `Jetpack_Options` across all versions of Jetpack with this class?

abelsonlive replied:
I don't know that for certain. I was following the pattern that was previously established in [`WC_Tracks`](https://github.com/woocommerce/woocommerce/blob/eda463d542a3206f4068ddbcdaab60cb81754838/plugins/woocommerce/includes/tracks/class-wc-tracks.php#L40) and [`WC_Email_New_Order`](https://github.com/woocommerce/woocommerce/blob/546ef267a92f45a5fee71e9b06349d3479846c3a/plugins/woocommerce/includes/emails/class-wc-email-new-order.php#L257).

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

abelsonlive commented on plugins/woocommerce/includes/class-wc-install.php:
```diff
@@ -1217,11 +1239,10 @@ private static function get_schema() {
 
 		$feature_controller = wc_get_container()->get( FeaturesController::class );
-		$hpos_enabled =
+		$hpos_enabled       =
```
linting change

abelsonlive commented on plugins/woocommerce/includes/class-wc-install.php:
```diff
@@ -1217,11 +1239,10 @@ private static function get_schema() {
-			self::should_enable_hpos_for_new_shop()
-		;
-		$hpos_table_schema = $hpos_enabled ? wc_get_container()->get( OrdersTableDataStore::class )->get_database_schema() : '';
+			self::should_enable_hpos_for_new_shop();
+		$hpos_table_schema  = $hpos_enabled ? wc_get_container()->get( OrdersTableDataStore::class )->get_database_schema() : '';
```
linting change.

louwie17 commented on plugins/woocommerce/includes/tracks/class-wc-tracks.php:
```diff
@@ -38,6 +38,7 @@ public static function get_blog_details( $user_id ) {
+				'store_id'       => get_option( 'woocommerce_store_id', null ),
 				'products_count' => self::get_products_count(),
 				'wc_version'     => WC()->version,
 			);
```
Just a note, on the line below, is that these `$blog_details` are being cached in a transient for a day.
Which means that the `store_id` won't be added until a day after they update.
This isn't a huge deal in my opinion, but drove me slightly crazy while trying to test if it was set correctly after update.

abelsonlive replied:
Yes, this was confusing to me, as well! Watch out for this when testing @deltaWhiskey...

nerrad replied:
Does the 24 hour day delay matter for this new option? If so, you could probably bust the cache if the id is not set yet?

louwie17 reviewed (approved):
This tested well @abelsonlive and the changes look great to me! Nice work!
I did leave one inline comment around the tracking data being cached in a transient for 1 day. So this may mean you won't see this data until a day after the release.

deltaWhiskey reviewed (approved):
Code looks good, and is working as expected.

nerrad commented on plugins/woocommerce/includes/tracks/class-wc-tracks.php:
```diff
@@ -38,6 +38,7 @@ public static function get_blog_details( $user_id ) {
 				'url'            => home_url(),
 				'blog_lang'      => get_user_locale( $user_id ),
 				'blog_id'        => class_exists( 'Jetpack_Options' ) ? Jetpack_Options::get_option( 'id' ) : null,
+				'store_id'       => get_option( 'woocommerce_store_id', null ),
```
Since there is a constant for this option name, should you use it? (This also applies to elsewhere the string is directly referenced).

abelsonlive replied:
Good point! I switched all references to `WC_INSTALL::STORE_ID_OPTION`

nerrad commented on plugins/woocommerce/tests/php/includes/class-wc-install-test.php:
```diff
@@ -164,4 +164,22 @@ public function test_delete_obsolete_notes_deletes_only_selected_notes() {
+	public function test_maybe_set_store_id() {
+
+		// simulate a store ID not being set.
+		delete_option( \WC_Install::STORE_ID_OPTION );
+		$store_id = \WC_Install::maybe_set_store_id();
+		$this->assertSame( $store_id, get_option( \WC_Install::STORE_ID_OPTION ) );
+		// uuid4 is 36 characters long.
+		$this->assertSame( 36, strlen( $store_id ) );
+
+		// simulate a store ID already being set.
+		$existing_store_id = \WC_Install::maybe_set_store_id();
+		$this->assertSame( $store_id, $existing_store_id );
+		// cleanup.
+		delete_option( \WC_Install::STORE_ID_OPTION );
```
As I mentioned above, you should be able to do this test without needing to get the actual Store ID. Simply set the ID, retrieve it via `get_option`, call `maybe_set_store_id` again and ensure it stays the same. 

Although deleting the option is needed to ensure a clean state for the test, I don't think you need to test that what is saved to the option is the returned store_id - it's testing WP internal implementation details for the `wp_option` API which is already covered by WP core tests.

abelsonlive replied:
Got it! I updated the tests to not rely on `maybe_set_store_id` returning a value and removed the check for whether the ID is in fact the same. I kept the check for 36 characters, though, because I want to ensure that the id is a valid `uuid4` string.

The second half of the test checks the state when the `store_id` has already been set.

abelsonlive committed:
fixes after PR review

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

abelsonlive committed:
update version

abelsonlive commented:
@nerrad Thank you so much for your review! I really appreciate it as this is my first WC PR and I'm also new to PHP/WP development. Getting an extra set of eyes on this is extremely helpful :)

I addressed all your comments and followed your suggested changes. Let me know if there's anything else you think I should fix!

abelsonlive committed:
fix linting error

nerrad reviewed (approved):
Changes look good. Noting that I _did not test this_ on the assumption the other reviewer tests are sufficient.

I did have a minor comment/question that I'll leave up to you to decide on.
